### ISSUE TITLE
Lint for unintentional assignment in `if`/`do`/`while`

### ISSUE THREAD

**Issue by dnfield** [dart-lang/linter#3026]
Similar to https://github.com/dart-lang/sdk/issues/47438

BAD:

```dart
bool foo = false;

// ...

if (foo = true) {}
else if (foo = false) {}

while (foo = true) {}  // or false

do {} while (foo = true); // or false
```

GOOD:

```dart
bool foo = false;

// ...

if (foo == true) {}

while (foo == true) {}

do {} while (foo == true);
```

These lints should apply if the literal `true` or `false` is used. I'm not sure if they should apply when a const expression that evaluates to true or false is used. I think they should not apply if some non-const expression is used, e.g. `while (keepGoing = checkIfIShouldKeepGoing())` would not trigger the lint.

**Comment by natebosch** [comment]
FWIW `?? false` is a more idiomatic pattern than `== true` for handling `bool?` which would make the GOOD examples

```dart
bool? foo = false;

// ...

if (foo ?? false) {}

while (foo ?? false) {}

do {} while (foo ?? false);
```

**Comment by jamesderlin** [comment]
I agree that `== true` and `== false` checks are a code smell and should be avoided.  However, see https://stackoverflow.com/q/70793299/ for another case where someone accidentally did `if (booleanVariable = anotherBooleanVariable)`, which perhaps is a bit more compelling (although also uncommon).
